From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM codedrop s390-02-08-october2005

  Description: qeth: Kernel Oops due to NULL pointer dereference
  Symptom:     kernel oops when trying to bring up a qeth device online
               the very first time.
  Problem:     When trying to set a qeth device online the very first time
               it could happen that hardware reports device or
               channel errors before we have allocated and initialized a
               net_device. When this happens qeth is starting a recovery
               process and when calling qeth_set_offline we try to use
               card->dev which is a NULL pointer .
  Solution:    Before using netif_carrier_off check first if card->dev is
               not NULL and if there is a need to call netif_carrier_off.
  Problem-ID:  20908
  Archs:       s390-31, s390-64

Signed-off-by: Jan Blunck <jblunck@suse.de>

--- linux-2.5/drivers/s390/net/qeth_main.c	4 Jan 2006 17:59:20 -0000	1.252.2.1
+++ linux-2.5/drivers/s390/net/qeth_main.c	2 Feb 2006 17:48:30 -0000	1.252.2.2
@@ -520,7 +520,8 @@
 	QETH_DBF_TEXT(setup, 3, "setoffl");
 	QETH_DBF_HEX(setup, 3, &card, sizeof(void *));
 	
-	netif_carrier_off(card->dev);
+	if (card->dev && netif_carrier_ok(card->dev))
+		netif_carrier_off(card->dev);
 	recover_flag = card->state;
 	if (qeth_stop_card(card, recovery_mode) == -ERESTARTSYS){
 		PRINT_WARN("Stopping card %s interrupted by user!\n",
@@ -1703,7 +1704,8 @@
 					   QETH_CARD_IFNAME(card),
 					   card->info.chpid);
 				card->lan_online = 0;
-				netif_carrier_off(card->dev);
+				if (card->dev && netif_carrier_ok(card->dev))
+					netif_carrier_off(card->dev);
 				return NULL;
 			case IPA_CMD_STARTLAN:
 				PRINT_INFO("Link reestablished on %s "

